design/pv-migration.md

```diff
@@ -0,0 +1,86 @@
+---
+title: Volume Migration for ZFS-LocalPV
+authors:
+  - "@pawanpraka1"
+owners:
+  - "@kmova"
+creation-date: 2021-05-21
+last-updated: 2021-05-21
+---
+
+# Volume Migration for ZFS-LocalPV
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Problem](#problem)
+* [Current Solution](#current-solution)
+* [Proposal](#proposal)
+    * [Keys Per ZPOOL](#keys-per-zpool)
+    * [Migrator](#migrator)
+    * [Workflow](#workflow)
+* [Implementation Plan](#implementation-plan)
+
+## Summary
+
+This is a design proposal to implement a feature for migrating a volume from one node to another node. This doc describes how we can move the persistence volumes and the application to the other node for ZFS-LocalPV CSI Driver. This design expects that the administrators will move the disks to the new node and will import the pool there as part of replacing the node. The goal of this design is volume and the pod using the volume should be moved to the new nodes. This design also assumes that admins are not having large number of ZFS POOLs configured on a node.
+
+## Problem
+
+The problem with the LocalPV is, it has the affinity set on the PersistenceVolume object. This will let k8s scheduler to schedule the pods to that node as data is there only. ZFS-LocalPV driver uses nodename to set the affinity which creates the problem here as if we are replacing a node, the node name will change and k8s scheduler will not be able to schedule the pods to the new node even if we have moved the disks there.
+
+## Current Solution
+
+Current solution depends on `openebs.io/nodeid` node label. Admin has to set the same label to the replaced node which allows the k8s scheduler to automatically schedule the old pods to this new node. We can read more about the solution [here](https://github.com/openebs/zfs-localpv/blob/master/docs/faq.md#8-how-to-migrate-pvs-to-the-new-node-in-case-old-node-is-not-accessible).
+
+The problem with the above approach is we can not move the volumes to any existing node as we can set only one label on the node for a given key. So, for any existing node in a cluster, label `openebs.io/nodeid` will already be set so, we can not update it to a new value as the pods already running on it will not able to get scheduled here.
+
+
+## Proposal
+
+### Keys Per ZPOOL
+
+We are proposing to have a key dedicated to ZFS POOL. This key will be used by the ZFS-LocalPV driver to set the label on the nodes where it is present. In this way we can allow the ZFS POOLs to move from any node to any other node as the key is tied to the ZFS POOL as opposed to keeping it per node. We are proposing to have a `guid.openebs.io/<pool-guid>=true` label on the node where the pool is present. Assuming admins do not have large number of pools on a node, there will be not much label set on a node.
```

Comment:
How about following the naming convention `guid.zfspool.openebs.io/<pool-guid>=true`. Open for discussion on a naming convention

Reply:
`guid.zfs.openebs.io/<pool-guid>=true`. can also be considered, since we already have `zfs.openebs.io`

Reply:
thinking of making it `guid.zpool.openebs.io/<pool-guid>=true` as it specifically says that it is the guid of zpool. Will wait for others to comment on this one.

Reply:
`guid.zpool.openebs.io/<pool-guid>=true` make more sense to me.

Reply:
`guid.zpool.openebs.io/<pool-guid>=true` is fine. 

But if we need to maintain consistency we can also use `zpool-guid.zfs.openebs.io/<pool-guid>=true`

Also, considering how the LVM will make use of this field can also help standardize the parameter.